third_party/java/arcs/build_defs/internal/plan.bzl

```diff
@@ -0,0 +1,83 @@
+"""Arcs Plan Generation Rules"""
+
+load("//third_party/java/arcs/build_defs/internal:kotlin.bzl", "arcs_kt_library")
+
+# Note: Once this is mature, it will replace arcs_kt_plan
+def arcs_kt_plan_2(name, package, arcs_sdk_deps, srcs = [], deps = [], visibility = None):
+    """Generates Plans Jar from protos.
+
+    Example:
+
+      ```
+      arcs_proto_plan(
+        name = "serialized_example",
+        src = "Example.arcs"
+      )
+
+      arcs_plan_generation(
+        name = "example_plan",
+        srcs = [":serialized_example"],
+        package = "com.my.example",
+      )
+      ```
+
+    Arcs:
+      name: name of created target
+      package: the package that all generated code will belong to (temporary, see b/161994250).
+      arcs_sdk_deps: build targets for the Arcs SDK to be included
+      deps: JVM dependencies for Jar
+      visibility: list of visibilities
+    """
+
+    gen_name = name + "_GeneratedPlan"
+
+    recipe2plan(
+        name = gen_name,
+        srcs = srcs,
+        package = package,
+    )
+
+    arcs_kt_library(
+        name = name,
+        srcs = [":" + gen_name],
+        platforms = ["jvm"],
+        visibility = visibility,
+        deps = arcs_sdk_deps + deps,
+    )
+
+def _recipe2plan_impl(ctx):
```

Comment:
This all looks good, but it's quite complex, and I suspect we will want to do this later on for other rules too. 

I have a PR out for review https://github.com/PolymerLabs/arcs/pull/5911 which introduces a `run_test` helper rule to invoke binaries with args for tests. This strikes me as completely equivalent to what you need here, except it's for tests. 

I'll see if I can put together a `test_binary` rule, shouldn't take long. I want to replace `run_in_repo` with that anyway, so it's not wasted work.

Don't block on me if I don't get it done in time, it's easy to refactor later on. Plus what you have here looks good anyway.

Reply:
I looked at the PR out for review, and I disagree. I feel like this rule is really simple for a bazel rule. It seems to me that it would be more complex to have one rule for everything. That would make adding custom providers per rule really hard, which is something that we would like to do soon.

Reply:
My point was that this rule doesn't need be a fully-fledged rule, it could be a really simple macro. (All it does it invoke a binary, it's not very complicated.) I don't really see how it would utilise providers, but it's possible I'm just not thinking hard enough about it.

Anyway, I'm happy for you to leave it as it is. It's a little upsetting that bazel makes it so hard to do something as simple as run a binary that it already built for you, but it is what is it...

Reply:
It's not taking advantage of any provider system in this PR, however, I have plans for this mechanism later.